lib/Sema/CSGen.cpp

```diff
@@ -1019,6 +1019,36 @@ namespace {
       return tv;
     }
 
+    Type addUnresolvedMemberChainConstraints(Expr *expr, Type resultTy,
+                                             unsigned additionalOptions = 0) {
+      // If this is a member chain hanging off of an UnresolvedMemberExpr,
+      // and we're at the last element of the chain, then the contextual type
+      // (represented with a new type variable) must equal the base type.
+      if (CS.isMemberChainTail(expr)) {
+        auto *chainBaseExpr = CS.getMemberChainBase(expr);
+        if (auto *UME = dyn_cast<UnresolvedMemberExpr>(chainBaseExpr)) {
```

Comment:
Should this be a `castTo<UnresolvedMemberExpr>` instead since base of the member chain is always an `UnresolvedMemberExpr`?

Reply:
I don't think so—`getMemberChainBase` works for all member chains, including, say `foo?.bar!.baz()`, so we do actually want to check here whether the member chain has a base which is an `UnresolvedMemberExpr`, as opposed to anything else that could be sitting at the base.

Perhaps this method would be better named something like `maybeAddUnresolvedMemberChainConstraints`, that just starts getting pretty long.

Reply:
I'm asking because `getMemberChainBase` is only used in combination with `isMemberChainTail` which hints at it being tailored for unresolved member based chains. It seems like all of this functionality could be localized to `ConstraintWalker` and chains like this determined upfront...